library/ssl_misc.h

```diff
@@ -769,16 +769,14 @@ struct mbedtls_ssl_handshake_params {
 #define SSL_XXDH_PSA_PEERKEY_SIZE MBEDTLS_PSA_MAX_FFDH_PUBKEY_LENGTH
 #endif
 
-#if (defined(PSA_WANT_ALG_ECDH) || defined(PSA_WANT_ALG_FFDH)) && \
-    (defined(MBEDTLS_USE_PSA_CRYPTO) || defined(MBEDTLS_SSL_PROTO_TLS1_3))
+#if defined(MBEDTLS_KEY_EXCHANGE_SOME_XXDH_ANY_PSA_ENABLED)
```

Comment:
Ok, this confirms my suspicion about the definition of `MBEDTLS_KEY_EXCHANGE_SOME_XXDH_1_2_ENABLED`: we don't want `MBEDTLS_KEY_EXCHANGE_SOME_DHE_ENABLED` to be enough to cause these fields to be present, because TLS 1.2 DHE will never use these.

These fields are used:
- by TLS 1.2 with `USE_PSA` to do ECDH or ECDHE;
- by TLS 1.3 to do ECDHE or FFDHE.
We want a guard that reflects exactly that, so that we don't waste RAM including those fields in builds that don't need them.

Suggestion:
```c
/* The handshake params structure has a set of fields called xxdh_psa which are used:
 * - by TLS 1.2 with `USE_PSA` to do ECDH or ECDHE;
 * - by TLS 1.3 to do ECDHE or FFDHE.
 * The following macros can be used to guard their declaration and use.
 */
#if defined(MBEDTLS_KEY_EXCHANGE_SOME_ECDH_OR_ECDHE_1_2_ENABLED) && \
    defined(MBEDTLS_USE_PSA_CRYPTO)
#define MBEDTLS_KEY_EXCHANGE_SOME_XXDH_PSA_1_2_ENABLED
#endif
#if defined(MBEDTLS_KEY_EXCHANGE_SOME_XXDH_PSA_1_2_ENABLED) || \
    defined(MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_SOME_EPHEMERAL_ENABLED)
#define MBEDTLS_KEY_EXCHANGE_SOME_XXDH_ANY_PSA_ENABLED
#endif
```

Reply:
I understand the point and I agree.
I only have a side question: using previous guards and enabling `PSA_WANT_ALG_FFDH` + `MBEDTLS_USE_PSA_CRYPTO`, we should end up with the same problem (waste of RAM) I introduced here with TLS 1.2 DHE. Or did I missed something?

Reply:
I think using the previous guards, if we enable `PSA_WANT_ALG_FFDH` and `USE_PSA_CRYPTO`, the fields will be included only if TLS 1.3 is enabled, which seems correct to me. Or are we talking about different things?

Reply:
For sure it's me that I'm missing something, but in the previous implementation I see 2 blocks ANDed together:

- (defined(PSA_WANT_ALG_ECDH) || defined(PSA_WANT_ALG_FFDH))
- (defined(MBEDTLS_USE_PSA_CRYPTO) || defined(MBEDTLS_SSL_PROTO_TLS1_3))

So we enable `PSA_WANT_ALG_FFDH` from the 1st and `MBEDTLS_USE_PSA_CRYPTO` from the 2nd we should enable the `xxd_psa` data, right? I mean there isn't a strict requirement for `MBEDTLS_SSL_PROTO_TLS1_3`...

Reply:
Ah, you mean the way it was before this PR? Yes indeed, it was including the fields in configurations they shouldn't. That was a much more general problem before this PR, and one of the things it's meant to improve.

Reply:
Semi-related note on complex preprocessor conditions: there's a known bug in the tool we use for indentation that it gets it wrong in multiline preprocessor conditions involving `defined`: the intentation of the continuation line is not aligned with the nesting of parentheses. So be careful when reading those.

Reply:
Yes, this is something I noticed and that often tricks me (luckily I added the `uncrustify` hook before `git commit`) ;)